Added an example code to demonstrate how to use deterministic image generation

cfgfung opened the pull request:
# What does this PR do?

This PR contains an example code to demonstrate how to use deterministic image generation using Gaudi and related optimized pipelines. It is based on the existing example code to modify.

# Result

First generation:
![image_1](https://github.com/huggingface/optimum-habana/assets/15855337/6945f357-c763-4287-bff4-2e778560e358)


Second generation:
![image_1](https://github.com/huggingface/optimum-habana/assets/15855337/10a2358e-7f9a-4726-83a2-e077393cb148)




As seen above, the generated images are similar for two different runs if the generation seed is fixed. It can be concluded that deterministic image generation works on Gaudi with just a simple patch.

# Benchmarking
The results are obtained by using the following command:
```
python test_script.py \
    --model_name_or_path runwayml/stable-diffusion-v1-5 \
    --prompts "An image of a squirrel in Picasso style" \
    --num_images_per_prompt 20 \
    --batch_size 4 \
    --image_save_dir /tmp/stable_diffusion_images \
    --use_habana \
    --use_hpu_graphs \
    --gaudi_config Habana/stable-diffusion \
    --bf16
```


| Type   | Generation Runtime | Generation Samples Per Second | Generation Steps Per Second|
| ------------- | ------------- | ------------- | ------------- | 
| Original | 140.8114 |  0.928 |  0.58 | 
| Patched  | 141.0243 |   0.929 |  0.58 | 

As seen above, there is almost no regression in terms of latency.

ssarkar2 commented on examples/stable-diffusion/deterministic_image_generation.py:
```diff
@@ -0,0 +1,240 @@
+
+    # Save images in the specified directory if not None and if they are in PIL format
+    if args.image_save_dir is not None:
+        import pdb; pdb.set_trace()
```
stray pdb

ssarkar2 reviewed (changes requested):
Needs license info on the top of the file.

ssarkar2 commented:
@cfgfung Could you please add a brief description in this PR of the difference between [deterministic_image_generation.py](https://github.com/huggingface/optimum-habana/blob/cd76a4e93cb7d60860956008086f9273475a6f2c/examples/stable-diffusion/deterministic_image_generation.py) and [text_to_image_generation.py](https://github.com/huggingface/optimum-habana/blob/cd76a4e93cb7d60860956008086f9273475a6f2c/examples/stable-diffusion/text_to_image_generation.py)

Would it suffice to add a flag `--deterministic` , call `generator = [torch.Generator(device="cpu").manual_seed(i) for i in range(args.num_images_per_prompt)]` if `deterministic` flag is set, and then pass it to the pipeline in text_to_image_generation.py? Why do we need an extra file (deterministic_image_generation.py) ?

cfgfung commented:
> @cfgfung Could you please add a brief description in this PR of the difference between [deterministic_image_generation.py](https://github.com/huggingface/optimum-habana/blob/cd76a4e93cb7d60860956008086f9273475a6f2c/examples/stable-diffusion/deterministic_image_generation.py) and [text_to_image_generation.py](https://github.com/huggingface/optimum-habana/blob/cd76a4e93cb7d60860956008086f9273475a6f2c/examples/stable-diffusion/text_to_image_generation.py)
> 
> Would it suffice to add a flag `--deterministic` , call `generator = [torch.Generator(device="cpu").manual_seed(i) for i in range(args.num_images_per_prompt)]` if `deterministic` flag is set, and then pass it to the pipeline in text_to_image_generation.py? Why do we need an extra file (deterministic_image_generation.py) ?

Hi. This deterministic image generation technique is used in other models/pipelines instead of just text_to_image_generation.py. Therefore I did not put it under the folder of stable_diffusion. This standalone example makes it much easier for users to see the fix required if they want to use it in their pipeline.

imangohari1 reviewed (changes requested):
@cfgfung 
Hi, 
Thank you for adding this example. 
I am requesting some logistical changes for now before we get down on details. 
Could you please apply them and update this PR?
Thank you. 
- please `sync` and `rebase` this branch with `main` opt-habana. Currently this branch is 110 commits behind main. 
- this example needs a `README.md` file. Please add the readme file we example of tests. Pls. add the performance obtained on gaudi2 within the readme file tests as well. 
- please run `make style` and push the changes. 
- these changes require at leaset 2 CI tests. Please follow the examples here: https://github.com/huggingface/optimum-habana/blob/main/tests/test_diffusers.py#L1727-L1801

imangohari1 commented:
@cfgfung 
Any update on this?

cfgfung commented:
> @cfgfung Any update on this?

Hi,

Thanks for the review.

I have synced and rebased the branch. 
Added a section to the original readme file. Did not put the performance data into the readme as the original readme does not have that.
Ran the `make style` command and cleaned some unused codes.
Also added 2 new CI tests and ran those tests in a separate file.

imangohari1 reviewed (changes requested):
Hi @cfgfung 
Thank you for the changes. 

I've gone through this PR and believe there are some work done here before we able to merge this as listed below. 
Please apply these and let's revisit afters. 
Thanks 
[0001-fea-updated-the-ci-with-g1.-added-use_cpu_seed-arg.-.patch](https://github.com/user-attachments/files/15978214/0001-fea-updated-the-ci-with-g1.-added-use_cpu_seed-arg.-.patch)

1.  I've made some changes to the current codes and provided you a patch (attached). please apply this with `git am 0001*` 
   - These changes are mostly on CI testing and adding arg option for both CPU/HPU seed generation
2.   After 1 is completed and pushed, I suggest we merge the `deterministic_image_generation.py` and `text_to_image_generation.py` together with optional argument that's needed to run deterministic img generation in that script. 
    - The added value here would be to avoid unnecessary duplication. 
    - The CI test and README files would need some minor name/arg changes after this. 

CI tests with proposed patch on 
G1 
```bash
[INFO|pipeline_stable_diffusion.py:410] 2024-06-25 20:24:50,685 >> 1 prompt(s) received, 20 generation(s) per prompt, 4 sample(s) per batch, 5 total batch(es).
100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 5/5 [02:18<00:00, 27.61s/it]
[INFO|pipeline_stable_diffusion.py:606] 2024-06-25 20:27:08,766 >> Speed metrics: {'generation_runtime': 138.0558, 'generation_samples_per_second': 0.302, 'generation_steps_per_second': 0.189}
Potential NSFW content was detected in one or more images. A black image will be returned instead. Try again with a different prompt and/or seed.
PASSED
Loading pipeline components...: 100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 7/7 [00:00<00:00, 14.36it/s]
[INFO|pipeline_utils.py:130] 2024-06-25 20:27:20,286 >> Enabled HPU graphs.
[INFO|configuration_utils.py:305] 2024-06-25 20:27:20,407 >> loading configuration file gaudi_config.json from cache at /root/.cache/huggingface/hub/models--Habana--stable-diffusion/snapshots/60ee357057ec90d2b183de22d0327ddd5d5a6db9/gaudi_config.json
[INFO|configuration_utils.py:358] 2024-06-25 20:27:20,407 >> GaudiConfig {
  "autocast_bf16_ops": null,
  "autocast_fp32_ops": null,
  "optimum_version": "1.20.0",
  "transformers_version": "4.40.2",
  "use_dynamic_shapes": false,
  "use_fused_adam": true,
  "use_fused_clip_norm": true,
  "use_torch_autocast": true
}

============================= HABANA PT BRIDGE CONFIGURATION ===========================
 PT_HPU_LAZY_MODE = 1
 PT_RECIPE_CACHE_PATH =
 PT_CACHE_FOLDER_DELETE = 0
 PT_HPU_RECIPE_CACHE_CONFIG =
 PT_HPU_MAX_COMPOUND_OP_SIZE = 9223372036854775807
 PT_HPU_LAZY_ACC_PAR_MODE = 1
 PT_HPU_ENABLE_REFINE_DYNAMIC_SHAPES = 0
---------------------------: System Configuration :---------------------------
Num CPU Cores : 144
CPU RAM       : 1056455400 KB
------------------------------------------------------------------------------
[INFO|pipeline_stable_diffusion.py:410] 2024-06-25 20:27:23,787 >> 1 prompt(s) received, 20 generation(s) per prompt, 4 sample(s) per batch, 5 total batch(es).
100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 5/5 [02:25<00:00, 29.03s/it]
[INFO|pipeline_stable_diffusion.py:606] 2024-06-25 20:29:48,967 >> Speed metrics: {'generation_runtime': 145.1524, 'generation_samples_per_second': 0.302, 'generation_steps_per_second': 0.189}
Potential NSFW content was detected in one or more images. A black image will be returned instead. Try again with a different prompt and/or seed.
Potential NSFW content was detected in one or more images. A black image will be returned instead. Try again with a different prompt and/or seed.
Potential NSFW content was detected in one or more images. A black image will be returned instead. Try again with a different prompt and/or seed.
PASSED
```
G2
```bash
[INFO|pipeline_stable_diffusion.py:410] 2024-06-25 20:12:45,394 >> 1 prompt(s) received, 20 generation(s) per prompt, 4 sample(s) per batch, 5 total batch(es).
100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 5/5 [02:16<00:00, 27.31s/it]
[INFO|pipeline_stable_diffusion.py:606] 2024-06-25 20:15:01,992 >> Speed metrics: {'generation_runtime': 136.5699, 'generation_samples_per_second': 1.008, 'generation_steps_per_second': 0.63}
Potential NSFW content was detected in one or more images. A black image will be returned instead. Try again with a different prompt and/or seed.
PASSED
Loading pipeline components...: 100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 7/7 [00:00<00:00, 13.27it/s]
[INFO|pipeline_utils.py:130] 2024-06-25 20:15:38,346 >> Enabled HPU graphs.
[INFO|configuration_utils.py:305] 2024-06-25 20:15:38,438 >> loading configuration file gaudi_config.json from cache at /root/.cache/huggingface/hub/models--Habana--stable-diffusion/snapshots/60ee357057ec90d2b183de22d0327ddd5d5a6db9/gaudi_config.json
[INFO|configuration_utils.py:358] 2024-06-25 20:15:38,438 >> GaudiConfig {
  "autocast_bf16_ops": null,
  "autocast_fp32_ops": null,
  "optimum_version": "1.20.0",
  "transformers_version": "4.40.2",
  "use_dynamic_shapes": false,
  "use_fused_adam": true,
  "use_fused_clip_norm": true,
  "use_torch_autocast": true
}

============================= HABANA PT BRIDGE CONFIGURATION ===========================
 PT_HPU_LAZY_MODE = 1
 PT_RECIPE_CACHE_PATH =
 PT_CACHE_FOLDER_DELETE = 0
 PT_HPU_RECIPE_CACHE_CONFIG =
 PT_HPU_MAX_COMPOUND_OP_SIZE = 9223372036854775807
 PT_HPU_LAZY_ACC_PAR_MODE = 1
 PT_HPU_ENABLE_REFINE_DYNAMIC_SHAPES = 0
---------------------------: System Configuration :---------------------------
Num CPU Cores : 160
CPU RAM       : 1056398828 KB
------------------------------------------------------------------------------
[INFO|pipeline_stable_diffusion.py:410] 2024-06-25 20:15:43,236 >> 1 prompt(s) received, 20 generation(s) per prompt, 4 sample(s) per batch, 5 total batch(es).
100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 5/5 [02:24<00:00, 28.81s/it]
[INFO|pipeline_stable_diffusion.py:606] 2024-06-25 20:18:07,328 >> Speed metrics: {'generation_runtime': 144.0675, 'generation_samples_per_second': 1.012, 'generation_steps_per_second': 0.633}
Potential NSFW content was detected in one or more images. A black image will be returned instead. Try again with a different prompt and/or seed.
PASSED
```

cfgfung commented:
> Hi @cfgfung Thank you for the changes.
> 
> I've gone through this PR and believe there are some work done here before we able to merge this as listed below. Please apply these and let's revisit afters. Thanks [0001-fea-updated-the-ci-with-g1.-added-use_cpu_seed-arg.-.patch](https://github.com/user-attachments/files/15978214/0001-fea-updated-the-ci-with-g1.-added-use_cpu_seed-arg.-.patch)
> 
> 1. I've made some changes to the current codes and provided you a patch (attached). please apply this with `git am 0001*`
> 
> * These changes are mostly on CI testing and adding arg option for both CPU/HPU seed generation
> 
> 2. After 1 is completed and pushed, I suggest we merge the `deterministic_image_generation.py` and `text_to_image_generation.py` together with optional argument that's needed to run deterministic img generation in that script.
>    - The added value here would be to avoid unnecessary duplication.
>    - The CI test and README files would need some minor name/arg changes after this.
> 
> CI tests with proposed patch on G1
> 
> ```shell
> [INFO|pipeline_stable_diffusion.py:410] 2024-06-25 20:24:50,685 >> 1 prompt(s) received, 20 generation(s) per prompt, 4 sample(s) per batch, 5 total batch(es).
> 100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 5/5 [02:18<00:00, 27.61s/it]
> [INFO|pipeline_stable_diffusion.py:606] 2024-06-25 20:27:08,766 >> Speed metrics: {'generation_runtime': 138.0558, 'generation_samples_per_second': 0.302, 'generation_steps_per_second': 0.189}
> Potential NSFW content was detected in one or more images. A black image will be returned instead. Try again with a different prompt and/or seed.
> PASSED
> Loading pipeline components...: 100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 7/7 [00:00<00:00, 14.36it/s]
> [INFO|pipeline_utils.py:130] 2024-06-25 20:27:20,286 >> Enabled HPU graphs.
> [INFO|configuration_utils.py:305] 2024-06-25 20:27:20,407 >> loading configuration file gaudi_config.json from cache at /root/.cache/huggingface/hub/models--Habana--stable-diffusion/snapshots/60ee357057ec90d2b183de22d0327ddd5d5a6db9/gaudi_config.json
> [INFO|configuration_utils.py:358] 2024-06-25 20:27:20,407 >> GaudiConfig {
>   "autocast_bf16_ops": null,
>   "autocast_fp32_ops": null,
>   "optimum_version": "1.20.0",
>   "transformers_version": "4.40.2",
>   "use_dynamic_shapes": false,
>   "use_fused_adam": true,
>   "use_fused_clip_norm": true,
>   "use_torch_autocast": true
> }
> 
> ============================= HABANA PT BRIDGE CONFIGURATION ===========================
>  PT_HPU_LAZY_MODE = 1
>  PT_RECIPE_CACHE_PATH =
>  PT_CACHE_FOLDER_DELETE = 0
>  PT_HPU_RECIPE_CACHE_CONFIG =
>  PT_HPU_MAX_COMPOUND_OP_SIZE = 9223372036854775807
>  PT_HPU_LAZY_ACC_PAR_MODE = 1
>  PT_HPU_ENABLE_REFINE_DYNAMIC_SHAPES = 0
> ---------------------------: System Configuration :---------------------------
> Num CPU Cores : 144
> CPU RAM       : 1056455400 KB
> ------------------------------------------------------------------------------
> [INFO|pipeline_stable_diffusion.py:410] 2024-06-25 20:27:23,787 >> 1 prompt(s) received, 20 generation(s) per prompt, 4 sample(s) per batch, 5 total batch(es).
> 100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 5/5 [02:25<00:00, 29.03s/it]
> [INFO|pipeline_stable_diffusion.py:606] 2024-06-25 20:29:48,967 >> Speed metrics: {'generation_runtime': 145.1524, 'generation_samples_per_second': 0.302, 'generation_steps_per_second': 0.189}
> Potential NSFW content was detected in one or more images. A black image will be returned instead. Try again with a different prompt and/or seed.
> Potential NSFW content was detected in one or more images. A black image will be returned instead. Try again with a different prompt and/or seed.
> Potential NSFW content was detected in one or more images. A black image will be returned instead. Try again with a different prompt and/or seed.
> PASSED
> ```
> 
> G2
> 
> ```shell
> [INFO|pipeline_stable_diffusion.py:410] 2024-06-25 20:12:45,394 >> 1 prompt(s) received, 20 generation(s) per prompt, 4 sample(s) per batch, 5 total batch(es).
> 100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 5/5 [02:16<00:00, 27.31s/it]
> [INFO|pipeline_stable_diffusion.py:606] 2024-06-25 20:15:01,992 >> Speed metrics: {'generation_runtime': 136.5699, 'generation_samples_per_second': 1.008, 'generation_steps_per_second': 0.63}
> Potential NSFW content was detected in one or more images. A black image will be returned instead. Try again with a different prompt and/or seed.
> PASSED
> Loading pipeline components...: 100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 7/7 [00:00<00:00, 13.27it/s]
> [INFO|pipeline_utils.py:130] 2024-06-25 20:15:38,346 >> Enabled HPU graphs.
> [INFO|configuration_utils.py:305] 2024-06-25 20:15:38,438 >> loading configuration file gaudi_config.json from cache at /root/.cache/huggingface/hub/models--Habana--stable-diffusion/snapshots/60ee357057ec90d2b183de22d0327ddd5d5a6db9/gaudi_config.json
> [INFO|configuration_utils.py:358] 2024-06-25 20:15:38,438 >> GaudiConfig {
>   "autocast_bf16_ops": null,
>   "autocast_fp32_ops": null,
>   "optimum_version": "1.20.0",
>   "transformers_version": "4.40.2",
>   "use_dynamic_shapes": false,
>   "use_fused_adam": true,
>   "use_fused_clip_norm": true,
>   "use_torch_autocast": true
> }
> 
> ============================= HABANA PT BRIDGE CONFIGURATION ===========================
>  PT_HPU_LAZY_MODE = 1
>  PT_RECIPE_CACHE_PATH =
>  PT_CACHE_FOLDER_DELETE = 0
>  PT_HPU_RECIPE_CACHE_CONFIG =
>  PT_HPU_MAX_COMPOUND_OP_SIZE = 9223372036854775807
>  PT_HPU_LAZY_ACC_PAR_MODE = 1
>  PT_HPU_ENABLE_REFINE_DYNAMIC_SHAPES = 0
> ---------------------------: System Configuration :---------------------------
> Num CPU Cores : 160
> CPU RAM       : 1056398828 KB
> ------------------------------------------------------------------------------
> [INFO|pipeline_stable_diffusion.py:410] 2024-06-25 20:15:43,236 >> 1 prompt(s) received, 20 generation(s) per prompt, 4 sample(s) per batch, 5 total batch(es).
> 100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 5/5 [02:24<00:00, 28.81s/it]
> [INFO|pipeline_stable_diffusion.py:606] 2024-06-25 20:18:07,328 >> Speed metrics: {'generation_runtime': 144.0675, 'generation_samples_per_second': 1.012, 'generation_steps_per_second': 0.633}
> Potential NSFW content was detected in one or more images. A black image will be returned instead. Try again with a different prompt and/or seed.
> PASSED
> ```


Hi,

I have merged the 001* patch. Added an argument (--deterministic) to the text_to_image_generation.py and removed the deterministic_image_generation.py.

Re-ran with style test:
![image](https://github.com/huggingface/optimum-habana/assets/15855337/ad103d38-b917-4bda-89c6-2be01783a96d)

Ran the unit test:
![image](https://github.com/huggingface/optimum-habana/assets/15855337/141dc4fe-0c54-449d-b68f-414ef8ff5bf2)

Also ran the example code:
![image](https://github.com/huggingface/optimum-habana/assets/15855337/77727495-1a58-4950-be8c-9fa30adff24f)

cfgfung commented:
@imangohari1 Would you please check the new changes?

imangohari1 commented:
> @imangohari1 Would you please check the new changes?

Hi @cfgfung 
Thanks for the update here. The changes are cleaner now. 
I am still trying to understand the changes were are making here vs the existing one. 

Could we get the same deterministic outputs by just running the existing code with a fixed `--seed`? 
Could you try running `--determinitic False` and pass a fix `-seed` and run say 3 back to back examples and compare the generated images?

cfgfung commented:
> > @imangohari1 Would you please check the new changes?
> 
> Hi @cfgfung Thanks for the update here. The changes are cleaner now. I am still trying to understand the changes were are making here vs the existing one.
> 
> Could we get the same deterministic outputs by just running the existing code with a fixed `--seed`? Could you try running `--determinitic False` and pass a fix `-seed` and run say 3 back to back examples and compare the generated images?

@imangohari1 
Yes, you can generate the same/very similar outputs by using --seed. It is another approach.

For --seed, it will affect the global context to have fixed seeds. Below is the snippet of the code:
```
...
random.seed(seed)
np.random.seed(seed)
torch.manual_seed(seed)
hpu_random.manual_seed_all(seed)
...
```
For --deterministic, it adds more flexibility instead of fixing the global context. It provides each pipeline with a standalone generator/seed. It means that if you run two pipelines in sequence, they will generate the same outputs if the generator is the same. This will not happen when you use global context to fix the randomness. Since this is an example code, I suggest keeping both approaches for users. They can refer to this reference to construct their applications/services.

dsocek commented on examples/stable-diffusion/README.md:
```diff
@@ -36,7 +36,8 @@ python text_to_image_generation.py \
     --gaudi_config Habana/stable-diffusion \
-    --bf16
+    --bf16 \
+    --deterministic
```
Please remove `--deterministic` for default recommended sample command line in README.

cfgfung replied:
Removed.

dsocek commented on examples/stable-diffusion/text_to_image_generation.py:
```diff
@@ -320,6 +325,12 @@ def main():
     if args.throughput_warmup_steps is not None:
         kwargs_call["throughput_warmup_steps"] = args.throughput_warmup_steps
 
+    if args.deterministic:
```
This argument usage does not make a good case.   I would strongly recommend to rename the argument to something like `--use_cpu_rng` for allowing user to switch between HPU-based and CPU-based random number generator.   Determinism is and should be controlled only by seed, not by hard-coding seeds in a loop.   If user needs deterministic behavior he or she should control that by seed.  This sample can be deterministic if same seed is used regardless of if its HPU or CPU random number generator.

regisss replied:
The issue here is that setting the generator on HPU doesn't ensure determinism right?
Does it slow the whole process down to set the generator on CPU?

cfgfung replied:
> This argument usage does not make a good case. I would strongly recommend to rename the argument to something like `--use_cpu_rng` for allowing user to switch between HPU-based and CPU-based random number generator. Determinism is and should be controlled only by seed, not by hard-coding seeds in a loop. If user needs deterministic behavior he or she should control that by seed. This sample can be deterministic if same seed is used regardless of if its HPU or CPU random number generator.

We cannot use HPU-based generator. It will have the following error:
```
RuntimeError: Device type HPU is not supported for torch.Generator() api.
```

It is  intended to be an example code for users. Users can pass in any seed/value, as long as they use CPU generator. The for loop here is just a demonstration. It is expected that users will put different values. 

If we use the same value and put  (i.e. --use_cpu_rng) into the generator list, all the outputs are the same. This contradicts the intention of ```args.num_images_per_prompt```. Therefore I added a simple heuristic (the "hard-coded" for loop) to maintain determinism and diversity.

cfgfung replied:
> The issue here is that setting the generator on HPU doesn't ensure determinism right? Does it slow the whole process down to set the generator on CPU?

Similar. HPU does not implement the torch.Generator() API. Therefore we have to use CPU.

It will not affect the performance/throughput.

dsocek replied:
> > This argument usage does not make a good case. I would strongly recommend to rename the argument to something like `--use_cpu_rng` for allowing user to switch between HPU-based and CPU-based random number generator. Determinism is and should be controlled only by seed, not by hard-coding seeds in a loop. If user needs deterministic behavior he or she should control that by seed. This sample can be deterministic if same seed is used regardless of if its HPU or CPU random number generator.
> 
> We cannot use HPU-based generator. It will have the following error:
> 
> ```
> RuntimeError: Device type HPU is not supported for torch.Generator() api.
> ```
> 
> It is intended to be an example code for users. Users can pass in any seed/value, as long as they use CPU generator. The for loop here is just a demonstration. It is expected that users will put different values.
> 
> If we use the same value and put (i.e. --use_cpu_rng) into the generator list, all the outputs are the same. This contradicts the intention of `args.num_images_per_prompt`. Therefore I added a simple heuristic (the "hard-coded" for loop) to maintain determinism and diversity.

Using generator via `set_seed()` from `optimum.habana.utils` should be a default.  You are adding a CPU based RNG generator for which you also can set seed.   Determinism is controlled by setting the same seed AND using same RNG.  So why not have `--seed` and `--use_cpu_rng` for these selections?

When user sets the same seed once, the RNG should spit out a deterministic sequence of pseudorandom numbers, so no need to set seed in a loop for each sample individually.

cfgfung replied:
OK. I will change the wordings from `--deterministic` to `--use_cpu_rng`. Also the CPU generator will take the values from `--seed`

imangohari1 replied:
thanks @cfgfung  but now the value of `manual_seed(args.seed)` is independent of `i`. 
Should we change that to `manual_seed(args.seed+i)`?

cfgfung replied:
@imangohari1 
Yes, it will be much better. It will be able to generate different images. I have added that.

dsocek commented on examples/stable-diffusion/text_to_image_generation.py:
```diff
@@ -238,6 +238,11 @@ def main():
         help="Path to lora id",
     )
+    parser.add_argument(
+        "--deterministic",
```
See comment below

cfgfung replied:
Wait for the discussion

dsocek commented on examples/stable-diffusion/text_to_image_generation.py:
```diff
@@ -320,6 +325,14 @@ def main():
+    if args.use_cpu_rng:
+        # Patch for the deterministic generation - Need to specify CPU as the torch generator
+        generator = [
+            torch.Generator(device="cpu").manual_seed(args.seed + i) for i in range(args.num_images_per_prompt)
```
This will fail when you have more than 1 prompt.   You would need to use `args.num_images_per_prompt * len(args.prompts)` instead.   But I am trying to understand why not just use `generator = torch.Generator(device="cpu").manual_seed(args.seed)` here?  Are we trying to demonstrate an example that you can set generator for each sample separately?

cfgfung replied:
Synced with @dsocek . It will be better to just pass one CPU generator instead of a list. I have modified the code.

dsocek commented on examples/stable-diffusion/text_to_image_generation.py:
```diff
@@ -327,9 +327,7 @@ def main():
-        generator = [
-            torch.Generator(device="cpu").manual_seed(args.seed + i) for i in range(args.num_images_per_prompt)
-        ]
+        generator = torch.Generator(device="cpu").manual_seed(args.seed)
```
Need to rebase

regisss commented on tests/test_diffusers.py:
```diff
@@ -2211,3 +2212,75 @@ def test_stable_video_diffusion_no_throughput_regression_bf16(self):
+                --use_hpu_graphs
+                --gaudi_config Habana/stable-diffusion
+                --bf16
+                --deterministic
```
```suggestion
                --use_cpu_rng
```

cfgfung replied:
Thanks for the catch 👍 

I have rebased the code to adapt the newly introduced multi-card inferencing. Also corrected the cmd argument `--deterministic` in the test file.

Tested the rebased codes using the new distributed inferencing and single card examples. They are all able to generate results. I have also run the test case for GaudiDeterministicImageGenerationTester. The result is shown below:

![image](https://github.com/user-attachments/assets/ef0129c0-3d3c-4a28-9d79-3c008f2060c7)

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/optimum-habana/pr_878). All of your documentation changes will be reflected on that endpoint. The docs are available until 30 days after the last update.
